pkg/dms/meta/topology.go

```diff
@@ -18,48 +18,209 @@ import (
-	DMServerConfigs struct {
-		Master map[string]interface{} `yaml:"dm-master"`
-		Worker map[string]interface{} `yaml:"dm-worker"`
-		Portal map[string]interface{} `yaml:"dm-portal"`
-	}
-
-	// DMTopologySpecification represents the specification of topology.yaml
-	DMTopologySpecification struct {
-		GlobalOptions    GlobalOptions      `yaml:"global,omitempty"`
-		MonitoredOptions MonitoredOptions   `yaml:"monitored,omitempty"`
-		ServerConfigs    DMServerConfigs    `yaml:"server_configs,omitempty"`
-		Masters          []MasterSpec       `yaml:"dm-master_servers"`
-		Workers          []WorkerSpec       `yaml:"dm-worker_servers"`
-		Portals          []PortalSpec       `yaml:"dm-portal_servers"`
-		Monitors         []PrometheusSpec   `yaml:"monitoring_servers"`
-		Grafana          []GrafanaSpec      `yaml:"grafana_servers,omitempty"`
-		Alertmanager     []AlertManagerSpec `yaml:"alertmanager_servers,omitempty"`
+	// InstanceSpec represent a instance specification
+	InstanceSpec interface {
+		Role() string
+		SSH() (string, int)
+		GetMainPort() int
+		IsImported() bool
+	}
+
+	// ResourceControl is used to control the system resource
+	// See: https://www.freedesktop.org/software/systemd/man/systemd.resource-control.html
+	ResourceControl = meta.ResourceControl
+
+	// GlobalOptions represents the global options for all groups in topology
+	// specification in topology.yaml
+	GlobalOptions = meta.GlobalOptions
+
+	// MonitoredOptions represents the monitored node configuration
+	MonitoredOptions = meta.MonitoredOptions
+
+	// DMSTopologySpecification represents the specification of topology.yaml
+	DMSTopologySpecification struct {
+		GlobalOptions    GlobalOptions    `yaml:"global,omitempty"`
+		MonitoredOptions MonitoredOptions `yaml:"monitored,omitempty"`
+		Job              JobSpec          `yaml:"job"`
 	}
 )
 
-// AllDMComponentNames contains the names of all dm components.
+// AllDMSComponentNames contains the names of all dm components.
 // should include all components in ComponentsByStartOrder
-func AllDMComponentNames() (roles []string) {
-	tp := &DMSpecification{}
+func AllDMSComponentNames() (roles []string) {
+	tp := &DMSSpecification{}
 	tp.IterComponent(func(c Component) {
 		roles = append(roles, c.Name())
 	})
 
 	return
 }
 
-// MasterSpec represents the Master topology specification in topology.yaml
-type MasterSpec struct {
+// JobSpec represents the data migration job that users want to do
+type JobSpec struct {
+	Action  string                 `yaml:"action"`
+	Type    string                 `yaml:"type"`
+	Name    string                 `yaml:"name,omitempty"`
+	Config  map[string]interface{} `yaml:"config,omitempty"`
+	Sources []SourceSpec           `yaml:"sources"`
+	Sink    SourceSpec             `yaml:"sink"`
+	Workers []WorkerSpec           `yaml:"workers"`
+	// job related specs, once job is stopped, these specs should stopped
+	DMMasters    []DMMasterSpec     `yaml:"dm-masters,omitempty"`
```

Comment:
Do we need the `dm-master` field?

Reply:
I distinguish specs into two types:
1. started with source, ended with source:
    For example, Dumpling, Lightning, Importer, DMWorker. When their related source is removed from the job, these components should be stopped.
2. started with job, ended with job:
    For example, DMMaster, Monitors. When one source is removed from the job, these components shouldn't be affected. But when the job is stopped, these components should be stopped.

For `dm-master`, I think one or two sources shouldn't affect `dm-master`, so I put it here.